### PR TITLE
Update bufr recipe

### PR DESCRIPTION
**Pull request by AlexanderRichert-NOAA**
This PR syncs up the bufr recipe from JCSDA spack with main Spack, plus a couple other tweaks (namely, removing the tests variant and using spack's built-in mechanism for running the unit tests instead).

### CODE REVIEW

**Review by edwardhartnett — APPROVED** [review]
Wow, the spack build of bufr is getting complex!

**Review by edwardhartnett — APPROVED** [review]

**Comment by alecbcs** [comment]
It looks like the archive tarball changed naming formats from v12.0.0 to v12.0.1.
```
alecbcs@laptop ~ % spack checksum bufr 12.0.1
==> Error: Could not find any remote versions for bufr
```

```
--- Old format --
https://github.com/NOAA-EMC/NCEPLIBS-bufr/archive/refs/tags/bufr_v12.0.0.tar.gz

--- New format ---
https://github.com/NOAA-EMC/NCEPLIBS-bufr/archive/refs/tags/v12.0.1.tar.gz
```

**Comment by jbathegit** [comment]
Yes, that was a planned change with v12.0.1, as noted in the discussion of https://github.com/NOAA-EMC/NCEPLIBS-bufr/issues/516

**Comment by AlexanderRichert-NOAA** [comment]
@alecbcs I think I accounted for this in my latest commit(s), can you git pull and check again?

File: var/spack/repos/builtin/packages/bufr/package.py [hunk 1]
```diff
@@ -11,37 +13,113 @@ class Bufr(CMakePackage):
...
+    def url_for_version(self, version):
+        if version >= Version("12.0.1"):
+            return (
+                "https://github.com/NOAA-EMC/NCEPLIBS-bufr/archive/refs/tags/v%s.tar.gz" % version
+            )
+        else:
+            return (
+                "https://github.com/NOAA-EMC/NCEPLIBS-bufr/archive/refs/tags/bufr_v%s.tar.gz"
+                % version
+            )
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/bufr/package.py]: Please make `url_for_version` the *first* method after the `depends_on` directives (so above `patch()`).

You can also simplify this method as follows, for example:

```suggestion
        pre = "bufr_" if version < Version("12.0.1") else ""
        return (
            f"https://github.com/NOAA-EMC/NCEPLIBS-bufr/archive/refs/tags/{pre}v
{version}.tar.gz"
        )
```

**Review by tldahlgren — CHANGES_REQUESTED** [review]
Confirmed all version sha256 with the new `url_for_version` but ask that it be moved.  Also suggesting a simplified implementation.

**Comment by AlexanderRichert-NOAA** [comment]
@spackbot fix style

**Review by alecbcs — DISMISSED** [review]
Thanks @AlexanderRichert-NOAA! I'd missed the update to `url_for_version`. This all looks good to me.

**Review by edwardhartnett — APPROVED** [review]

File: var/spack/repos/builtin/packages/bufr/package.py [hunk 2]
```diff
@@ -11,37 +13,108 @@ class Bufr(CMakePackage):
...
-        # Bufr has _DA (dynamic allocation) libs in versions <= 11.5.0
-        if self.spec.satisfies("@:11.5.0"):
-            da_lib = find_libraries(
-                libname + "_DA", root=self.prefix, shared=False, recursive=True
-            )
-            env.set(lib_envname + "_DA", da_lib[0])
-            env.set(inc_envname + "_DA", include_dir)
+        if self.spec.satisfies("+python"):
+            pyver = self.spec["python"].version.up_to(2)
+            pydir = os.path.join(os.path.dirname(lib[0]), "python%s/site-packages" % pyver)
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/bufr/package.py]: You're already using the path join so why not include the last directory node?  You can also use an f-string.

```suggestion
            pydir = join_path(os.path.dirname(lib[0]), f"python{pyver}", "site-packages")
```

File: var/spack/repos/builtin/packages/bufr/package.py [hunk 3]
```diff
@@ -11,37 +13,108 @@ class Bufr(CMakePackage):
...
+    # Need to make the lines shorter at least on some systems
+    def patch(self):
+        with when("@:11.7.1"):
+            filter_file("_lenslmax 120", "_lenslmax 60", "CMakeLists.txt")
+
+    def cmake_args(self):
+        args = [
+            self.define_from_variant("ENABLE_PYTHON", "python"),
+            self.define_from_variant("BUILD_SHARED_LIBS", "shared"),
+            self.define("BUILD_TESTS", self.run_tests),
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/bufr/package.py:67]: So I understand, you only want BUILD_TESTS enabled when the installer runs `spack install --test=root`.  Is that correct?

**Review by tldahlgren — DISMISSED** [review]
Confirmed the branch and all version sha256 (since adding `url_for_version`).  I noticed inconsistent use of f-strings in the additions and a path join with an explicit os separator.

Have a couple of questions.

**Comment by tldahlgren** [comment]
Any chance any of the patches can be upstreamed to the source repository?

**Comment by tldahlgren** [comment]
I'm able to successfully build (and execute post-install tests) using `spack -v install --test=root bufr`.

File: var/spack/repos/builtin/packages/bufr/package.py (hunk 3 above)
Comment (AlexanderRichert-NOAA) [reply var/spack/repos/builtin/packages/bufr/package.py:67]: Correct, but just to check-- @climbfuji it looks like you added the "tests" variant, do you have any objection to doing it just through `spack install --test`? Currently it doesn't look like it actually installs the tests, it just builds them and then they'll get wiped out unless using `--keep-stage`.

**Comment by AlexanderRichert-NOAA** [comment]
> Any chance any of the patches can be upstreamed to the source repository?

I just added a version range for the python one since that was just recently fixed, so all these fixes exist in at least the latest version of the library.

File: var/spack/repos/builtin/packages/bufr/package.py (hunk 3 above)
Comment (climbfuji) [reply var/spack/repos/builtin/packages/bufr/package.py:67]: No objections at all, thanks for asking!

Comment (tldahlgren) [reply var/spack/repos/builtin/packages/bufr/package.py:67]: That makes sense based on the nature of the check.  IF you have tests you want to cache from the build for use as stand-alone tests (days to weeks post-build), you'll want to check out the docs at https://spack.readthedocs.io/en/latest/packaging_guide.html#saving-build-time-files.

**Comment by tldahlgren** [comment]
@adamjstewart Do you want to take a look at the python aspects of this PR since the package is now extending `python`?

**Comment by tldahlgren** [comment]
FWIW. `spack -v install --test=root bufr +python` fails for me:

```
...
  >> 2261    The following tests FAILED:
     2262    	  1 - test_pyncepbufr_checkpoint (Failed)
     2263    	  2 - test_pyncepbufr_gps (Failed)
     2264    	  3 - test_pyncepbufr_prepbufr (Failed)
     2265    	  4 - test_pyncepbufr_rad (Failed)
     2266    	  5 - test_pyncepbufr_satwnd (Failed)
     2267    	  6 - test_pyncepbufr_write (Failed)
     2268    	  7 - test_pyncepbufr_test (Failed)
     2269    Errors while running CTest
     2270    Output from these tests are in: /tmp/dahlgren/spack-stage/spack-st
             age-bufr-12.0.1-p6rytfmhpqtfbuylfd3qsk537djzgkip/spack-build-p6ryt
             fm/Testing/Temporary/LastTest.log
     2271    Use "--rerun-failed --output-on-failure" to re-run the failed case
             s verbosely.
  >> 2272    make: *** [Makefile:74: test] Error 8
  ...
  ```
  
  With test details:
  
```
...
Start testing: Sep 25 15:15 PDT
----------------------------------------------------------
1/97 Testing: test_pyncepbufr_checkpoint
1/97 Test: test_pyncepbufr_checkpoint
Command: "/usr/WS1/dahlgren/releases/spack/opt/spack/linux-rhel8-broadwell/gcc-10.3.1/python-3.10.10-bmsjbb3kgcn5e7tgey5w7rlps2byhbhx/bin/python3.10" "/var/tmp/dahlgren/spack-stage/spack-stage-bufr-12.0.1-p6rytfmhpqtfbuylfd3qsk537djzgkip/spack-src/python/test/test_checkpoint.py"
Directory: /tmp/dahlgren/spack-stage/spack-stage-bufr-12.0.1-p6rytfmhpqtfbuylfd3qsk537djzgkip/spack-build-p6rytfm/test/testfiles
"test_pyncepbufr_checkpoint" start time: Sep 25 15:15 PDT
Output:
----------------------------------------------------------
Traceback (most recent call last):
  File "/var/tmp/dahlgren/spack-stage/spack-stage-bufr-12.0.1-p6rytfmhpqtfbuylfd3qsk537djzgkip/spack-src/python/test/test_checkpoint.py", line 2, in <module>
    import ncepbufr
ModuleNotFoundError: No module named 'ncepbufr'
<end of output>
Test time =   0.08 sec
----------------------------------------------------------
Test Failed.
"test_pyncepbufr_checkpoint" end time: Sep 25 15:15 PDT
"test_pyncepbufr_checkpoint" time elapsed: 00:00:00
----------------------------------------------------------

2/97 Testing: test_pyncepbufr_gps
2/97 Test: test_pyncepbufr_gps
Command: "/usr/WS1/dahlgren/releases/spack/opt/spack/linux-rhel8-broadwell/gcc-10.3.1/python-3.10.10-bmsjbb3kgcn5e7tgey5w7rlps2byhbhx/bin/python3.10" "/var/tmp/dahlgren/spack-stage/spack-stage-bufr-12.0.1-p6rytfmhpqtfbuylfd3qsk537djzgkip/spack-src/python/test/test_gps.py"
Directory: /tmp/dahlgren/spack-stage/spack-stage-bufr-12.0.1-p6rytfmhpqtfbuylfd3qsk537djzgkip/spack-build-p6rytfm/test/testfiles
"test_pyncepbufr_gps" start time: Sep 25 15:15 PDT
Output:
----------------------------------------------------------
Traceback (most recent call last):
  File "/var/tmp/dahlgren/spack-stage/spack-stage-bufr-12.0.1-p6rytfmhpqtfbuylfd3qsk537djzgkip/spack-src/python/test/test_gps.py", line 2, in <module>
    import ncepbufr
ModuleNotFoundError: No module named 'ncepbufr'
<end of output>
Test time =   0.04 sec
----------------------------------------------------------
Test Failed.
"test_pyncepbufr_gps" end time: Sep 25 15:15 PDT
"test_pyncepbufr_gps" time elapsed: 00:00:00
----------------------------------------------------------

3/97 Testing: test_pyncepbufr_prepbufr
3/97 Test: test_pyncepbufr_prepbufr
Command: "/usr/WS1/dahlgren/releases/spack/opt/spack/linux-rhel8-broadwell/gcc-10.3.1/python-3.10.10-bmsjbb3kgcn5e7tgey5w7rlps2byhbhx/bin/python3.10" "/var/tmp/dahlgren/spack-stage/spack-stage-bufr-12.0.1-p6rytfmhpqtfbuylfd3qsk537djzgkip/spack-src/python/test/test_prepbufr.py"
Directory: /tmp/dahlgren/spack-stage/spack-stage-bufr-12.0.1-p6rytfmhpqtfbuylfd3qsk537djzgkip/spack-build-p6rytfm/test/testfiles
"test_pyncepbufr_prepbufr" start time: Sep 25 15:15 PDT
Output:
----------------------------------------------------------
Traceback (most recent call last):
  File "/var/tmp/dahlgren/spack-stage/spack-stage-bufr-12.0.1-p6rytfmhpqtfbuylfd3qsk537djzgkip/spack-src/python/test/test_prepbufr.py", line 2, in <module>
    import ncepbufr
ModuleNotFoundError: No module named 'ncepbufr'
<end of output>
Test time =   0.04 sec
----------------------------------------------------------
Test Failed.
"test_pyncepbufr_prepbufr" end time: Sep 25 15:15 PDT
"test_pyncepbufr_prepbufr" time elapsed: 00:00:00
----------------------------------------------------------

4/97 Testing: test_pyncepbufr_rad
4/97 Test: test_pyncepbufr_rad
Command: "/usr/WS1/dahlgren/releases/spack/opt/spack/linux-rhel8-broadwell/gcc-10.3.1/python-3.10.10-bmsjbb3kgcn5e7tgey5w7rlps2byhbhx/bin/python3.10" "/var/tmp/dahlgren/spack-stage/spack-stage-bufr-12.0.1-p6rytfmhpqtfbuylfd3qsk537djzgkip/spack-src/python/test/test_rad.py"
Directory: /tmp/dahlgren/spack-stage/spack-stage-bufr-12.0.1-p6rytfmhpqtfbuylfd3qsk537djzgkip/spack-build-p6rytfm/test/testfiles
"test_pyncepbufr_rad" start time: Sep 25 15:15 PDT
Output:
----------------------------------------------------------
Traceback (most recent call last):
  File "/var/tmp/dahlgren/spack-stage/spack-stage-bufr-12.0.1-p6rytfmhpqtfbuylfd3qsk537djzgkip/spack-src/python/test/test_rad.py", line 2, in <module>
    import ncepbufr
ModuleNotFoundError: No module named 'ncepbufr'
<end of output>
Test time =   0.04 sec
----------------------------------------------------------
Test Failed.
"test_pyncepbufr_rad" end time: Sep 25 15:15 PDT
"test_pyncepbufr_rad" time elapsed: 00:00:00
----------------------------------------------------------

5/97 Testing: test_pyncepbufr_satwnd
5/97 Test: test_pyncepbufr_satwnd
Command: "/usr/WS1/dahlgren/releases/spack/opt/spack/linux-rhel8-broadwell/gcc-10.3.1/python-3.10.10-bmsjbb3kgcn5e7tgey5w7rlps2byhbhx/bin/python3.10" "/var/tmp/dahlgren/spack-stage/spack-stage-bufr-12.0.1-p6rytfmhpqtfbuylfd3qsk537djzgkip/spack-src/python/test/test_satwnd.py"
Directory: /tmp/dahlgren/spack-stage/spack-stage-bufr-12.0.1-p6rytfmhpqtfbuylfd3qsk537djzgkip/spack-build-p6rytfm/test/testfiles
"test_pyncepbufr_satwnd" start time: Sep 25 15:15 PDT
Output:
----------------------------------------------------------
Traceback (most recent call last):
  File "/var/tmp/dahlgren/spack-stage/spack-stage-bufr-12.0.1-p6rytfmhpqtfbuylfd3qsk537djzgkip/spack-src/python/test/test_satwnd.py", line 2, in <module>
    import ncepbufr
ModuleNotFoundError: No module named 'ncepbufr'
<end of output>
Test time =   0.04 sec
----------------------------------------------------------
Test Failed.
"test_pyncepbufr_satwnd" end time: Sep 25 15:15 PDT
"test_pyncepbufr_satwnd" time elapsed: 00:00:00
----------------------------------------------------------

6/97 Testing: test_pyncepbufr_write
6/97 Test: test_pyncepbufr_write
Command: "/usr/WS1/dahlgren/releases/spack/opt/spack/linux-rhel8-broadwell/gcc-10.3.1/python-3.10.10-bmsjbb3kgcn5e7tgey5w7rlps2byhbhx/bin/python3.10" "/var/tmp/dahlgren/spack-stage/spack-stage-bufr-12.0.1-p6rytfmhpqtfbuylfd3qsk537djzgkip/spack-src/python/test/test_write.py"
Directory: /tmp/dahlgren/spack-stage/spack-stage-bufr-12.0.1-p6rytfmhpqtfbuylfd3qsk537djzgkip/spack-build-p6rytfm/test/testfiles
"test_pyncepbufr_write" start time: Sep 25 15:15 PDT
Output:
----------------------------------------------------------
Traceback (most recent call last):
  File "/var/tmp/dahlgren/spack-stage/spack-stage-bufr-12.0.1-p6rytfmhpqtfbuylfd3qsk537djzgkip/spack-src/python/test/test_write.py", line 2, in <module>
    import ncepbufr
ModuleNotFoundError: No module named 'ncepbufr'
<end of output>
Test time =   0.05 sec
----------------------------------------------------------
Test Failed.
"test_pyncepbufr_write" end time: Sep 25 15:15 PDT
"test_pyncepbufr_write" time elapsed: 00:00:00
----------------------------------------------------------

7/97 Testing: test_pyncepbufr_test
7/97 Test: test_pyncepbufr_test
Command: "/usr/WS1/dahlgren/releases/spack/opt/spack/linux-rhel8-broadwell/gcc-10.3.1/python-3.10.10-bmsjbb3kgcn5e7tgey5w7rlps2byhbhx/bin/python3.10" "/var/tmp/dahlgren/spack-stage/spack-stage-bufr-12.0.1-p6rytfmhpqtfbuylfd3qsk537djzgkip/spack-src/python/test/test.py"
Directory: /tmp/dahlgren/spack-stage/spack-stage-bufr-12.0.1-p6rytfmhpqtfbuylfd3qsk537djzgkip/spack-build-p6rytfm/test/testfiles
"test_pyncepbufr_test" start time: Sep 25 15:15 PDT
Output:
----------------------------------------------------------
Traceback (most recent call last):
  File "/var/tmp/dahlgren/spack-stage/spack-stage-bufr-12.0.1-p6rytfmhpqtfbuylfd3qsk537djzgkip/spack-src/python/test/test.py", line 2, in <module>
    import ncepbufr
ModuleNotFoundError: No module named 'ncepbufr'
<end of output>
Test time =   0.04 sec
----------------------------------------------------------
Test Failed.
"test_pyncepbufr_test" end time: Sep 25 15:15 PDT
"test_pyncepbufr_test" time elapsed: 00:00:00
----------------------------------------------------------
...
```

**Comment by AlexanderRichert-NOAA** [comment]
@tldahlgren thanks I'll take a look, I wonder if it's a PYTHONPATH issue...

**Comment by AlexanderRichert-NOAA** [comment]
It looks like the problem is that the cmake config for the python component is built with the assumption that the user will run make install first. Since I don't think there's a way to change the order in spack, I think we might need to just disable testing and put in an issue for the library. @climbfuji thoughts? @tldahlgren any chance there's a way to either change the order (run 'make install' before 'make test') or set up a conflict so that +python can't be used if testing is enabled?

**Comment by climbfuji** [comment]
I'm fine disabling the testing, no problem.

**Comment by AlexanderRichert-NOAA** [comment]
For now I'm removing everything test related, then we can circle back after we sort things out upstream.

**Review by edwardhartnett — APPROVED** [review]

File: var/spack/repos/builtin/packages/bufr/package.py [hunk 4]
```diff
@@ -11,37 +13,103 @@ class Bufr(CMakePackage):
...
+    patch("python-version.patch", when="@11.5:12.0.0 +python")
+
+    variant("python", default=False, description="Enable Python interface?")
+    variant("shared", default=True, description="Build shared libraries", when="@11.5:")
+
+    extends("python", when="+python")
+
+    depends_on("python@3:", type=("build", "run"), when="+python")
+    depends_on("py-setuptools", type="build", when="+python")
+    depends_on("py-numpy", type="build", when="+python")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/bufr/package.py]: Probably needed at run-time too

Comment (AlexanderRichert-NOAA) [reply var/spack/repos/builtin/packages/bufr/package.py]: As far as I can tell, the ncepbufr python module does not need numpy at runtime, rather, numpy is just getting used in the setup.py.

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/bufr/package.py]: I see a few other places it's imported:
```console
> grep -R numpy
python/test/test.py:import numpy as np
python/test/test_write.py:import numpy as np
python/test/Python_tutorial_bufr.ipynb:    "import numpy as np\n",
python/utils/filterbyreceipt_fromsubset.py:import numpy as np
python/utils/filterbyreceipt_frommsg.py:import numpy as np
python/utils/prepbufr_rewrite:import numpy as np
python/utils/prepbufr2nc:import numpy as np
python/utils/nc2prepbufr:import numpy as np
python/utils/check_receipt_times.py:import numpy as np
python/ncepbufr/__init__.py:import numpy as np
```

Comment (AlexanderRichert-NOAA) [reply var/spack/repos/builtin/packages/bufr/package.py]: Ah right, thanks. Removing `type="build"`.

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/bufr/package.py]: The default type is build/link, it should be build/run

Comment (AlexanderRichert-NOAA) [reply var/spack/repos/builtin/packages/bufr/package.py]: Got it

**Review by adamjstewart — DISMISSED** [review]
Python bits look good now. No one has raised any issues with other parts of the code so I think it's safe to merge.

**Comment by adamjstewart** [comment]
@spackbot run pipeline

**Comment by tldahlgren** [comment]
> @tldahlgren any chance there's a way to either change the order (run 'make install' before 'make test') or set up a conflict so that +python can't be used if testing is enabled?

Since it is a `CMakePackage`, the `check` method is "reserved" for use post-build (see the table at https://spack.readthedocs.io/en/latest/packaging_guide.html#adding-installation-phase-tests).

If you want to add a post-install test, the docs mentioned above illustrates how to add a post-install method.

**Comment by AlexanderRichert-NOAA** [comment]
Thanks @tldahlgren -- it turns out the existing python testing is pretty dysfunctional anyway, so for now I'm just disabling testing when '+python' is enabled. This will be [fixed](https://github.com/NOAA-EMC/NCEPLIBS-bufr/pull/524) in the next release, so once 12.0.2 is out we can enable testing for the '+python' case.

**Review by tldahlgren — APPROVED** [review]
Still builds for me as expected with and without `python`.

**Pull request merged**